service/src/main/java/bio/terra/policy/db/PaoDao.java

```diff
@@ -200,6 +202,37 @@ public List<Pao> getPaos(List<UUID> objectIdList) {
     return paoList;
   }
 
+  /**
+   * Recursive function used by the delete PAO process. When a PAO is removed from the graph, we
+   * recursively inspect its sources to see if they now have no dependents. If they do not, then
+   * those PAOs can also be removed if they were previously flagged as deleted. And so on,
+   * recursively.
+   *
+   * @param pao the current PAO being evaluated.
+   * @param paosToRemove a set of PAO UUIDs for PAOs that should be removed from the graph after the
+   *     walk is performed.
+   */
+  private void walkRemovePao(DbPao pao, HashMap<UUID, DbPao> paosToRemove) {
```

Comment:
This is a difficult algorithm to follow, because the state of the list depends on the order of processing.
I _think_ this DAG does not work:
```
   A   B
   |\ /
   | v
   | C*
    \| \
     v  |
     D* |
     |  |
     |  |
      \ |
       v
       E*
```

- A on list
- A recurse to D
- D on list
- D recurse to E
- E on list
- E dep check: C not on list; remove E from list; return
- D dep check: C not on list; remove D from list; return
- A recurse to C
- C on list
- C recurse to D
- D on list
- D recurse to E
- E on list
- E dep check: C and D on list; leave E on list; return
- D dep check: C and A on list; leave D on list; return
- C recurse to E
- E on list
- E dep check: C and D on list; leave E on list; return
- C dep check: B not on list; remove C from list; return
- A dep check: none; return

Incorrectly leaves D and E on the list to be removed. Maybe... if I followed this correctly.

I don't have a solution off the top of my head. One idea is to do a full dependency query for dependents - maybe as a recursive SQL query. That way multi-path problems like the above would not be a factor. But that is expensive.
A second idea is some kind of two pass algorithm, building state across the subtree and then evaluating who can be deleted; that is, removing the ordering problem.